src/handlers/csv-extract/handler.test.ts

```diff
@@ -0,0 +1,310 @@
+    await expect(handler(validEvent)).rejects.toThrowError(
+      "Destination folder"
+    );
+  });
+
+  test("should throw error if event record does not have a JSON serialisable body", async () => {
+    const eventWithInvalidRecordBody = {
+      Records: [
+        {
+          body: "{",
+          messageId: "given message ID",
+        },
+      ],
+    };
+
+    const result = await handler(eventWithInvalidRecordBody as SQSEvent);
+
+    expect(result).toEqual({
+      batchItemFailures: [{ itemIdentifier: "given message ID" }],
+    });
+  });
+
+  test("should throw error if event record has body that does not serialise to object", async () => {
+    const eventWithInvalidRecordBody = {
+      Records: [
+        {
+          body: "1234",
+          messageId: "given message ID",
+        },
+      ],
+    };
+
+    const result = await handler(eventWithInvalidRecordBody as SQSEvent);
+
+    expect(result).toEqual({
+      batchItemFailures: [{ itemIdentifier: "given message ID" }],
+    });
+  });
+
+  test("should throw error if event record has invalid S3 event record in body", async () => {
+    const eventWithInvalidRecordBody = {
+      Records: [
+        {
+          body: JSON.stringify({
+            Records: [
+              {
+                s3: {
+                  bucket: {
+                    name: "some bucket name",
+                  },
+                  object: {
+                    key: 123412, // Invalid key because it is not a string
+                  },
+                },
+              },
+            ],
+          }),
+          messageId: "given message ID",
+        },
+      ],
+    };
+
+    const result = await handler(eventWithInvalidRecordBody as SQSEvent);
+
+    expect(result).toEqual({
+      batchItemFailures: [{ itemIdentifier: "given message ID" }],
+    });
+  });
+
+  test("should throw error with event record that has a valid S3 event in body with no vendor ID folder", async () => {
+    const givenEventWithNoFolder = {
+      Records: [
+        {
+          body: JSON.stringify({
+            Records: [
+              {
+                s3: {
+                  bucket: {
+                    name: "given bucket name",
+                  },
+                  object: {
+                    key: "given-file-path-with-no-folder",
+                  },
+                },
+              },
+            ],
+          }),
+          messageId: "given message ID",
+        },
+      ],
+    };
+
+    const result = await handler(givenEventWithNoFolder as SQSEvent);
+
+    expect(result).toEqual({
+      batchItemFailures: [{ itemIdentifier: "given message ID" }],
+    });
+  });
+
+  test("should throw error when given a valid bucket and key but fetchS3 throws error when fetching invoice", async () => {
+    const mockedErrorText = "mocked fetchS3 error";
+    const mockedError = new Error(mockedErrorText);
+    mockedFetchS3.mockRejectedValue(mockedError);
```

Comment:
Can we put in either `expect(mockedGetVendorServiceConfigRows).not.toBeCalled();` or `expect(mockedGetVendorServiceConfigRows).not.toBeCalled();` in all these unhappy path tests, just as a little assurance that they're failing for the right reason?

Reply:
Added the above and an expectHandlerFailure() function.